### PR TITLE
OpenSfM Update + Rolling Shutter correction

### PR DESCRIPTION
**Pull request by pierotofy**
Updates OpenSfM, ceres.

WIP for rolling shutter.

### CODE REVIEW

**Comment by smathermather** [comment]
I'm committed to not cursing, even in joy to pull requests, so I'll just say: wow.

File: opendm/rollingshutter.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    'dji fc300x': 33, # Phantom 3 Professional
+
+    'dji fc330': 33, # Phantom 4
+    'dji fc6310': 33, # Phantom 4 Professional
+
+    'dji fc350': 30, # Inspire 1
+
+    'gopro hero4 black': 30 # GoPro Hero 4 Black
+
+    # Help us add more! Open a pull request.
```
Comment (nyalldawson) [inline opendm/rollingshutter.py]: Any clues on how to determine this for a different model?

Comment (pierotofy) [reply opendm/rollingshutter.py]: We're trying to figure that out: https://community.opendronemap.org/t/phantom-4-photo-request-rolling-shutter/11777

Comment (nyalldawson) [reply opendm/rollingshutter.py]: Thanks @pierotofy -- this is very exciting!

**Comment by pierotofy** [comment]
Added camera velocities estimation / reading. We read velocity values from EXIF/XMP if available, or estimate them from camera poses.

```
2022-06-13 18:41:45,171 INFO: DJI_0151.JPG (+6.10,+0.20,-0.10) m/s
2022-06-13 18:41:45,171 INFO: DJI_0152.JPG (+6.30,+0.00,-0.10) m/s
2022-06-13 18:41:45,171 INFO: DJI_0153.JPG (+6.30,+0.00,+0.00) m/s
2022-06-13 18:41:45,171 INFO: DJI_0154.JPG (+6.30,+0.00,+0.00) m/s
2022-06-13 18:41:45,171 INFO: DJI_0155.JPG (+6.30,+0.00,+0.00) m/s
2022-06-13 18:41:45,171 INFO: DJI_0156.JPG (+6.30,+0.00,+0.00) m/s
2022-06-13 18:41:45,171 INFO: DJI_0157.JPG (+6.30,+0.00,+0.00) m/s
2022-06-13 18:41:45,171 INFO: DJI_0158.JPG (+6.40,+0.00,+0.00) m/s
```

**Comment by pierotofy** [comment]
The next steps are to:

1. Triangulate points from poses + observations.
2. For each 3D point, reproject the point back to image space --> reprojected observations
    * Calculate the error (reprojected observation - observation)
3. For each 3D point, reproject the point back to image space *using a corrected camera pose* --> adjusted reprojected observations
    * Calculate the corrected camera pose using the formula  image
    * `Ci` --> Camera origin (meters)
    * `Vuav` --> Velocity (meters / second)
    * `T` --> rolling shutter readout time (seconds)
    * `l` --> row number of observation (pixels)
    * `L` --> image height (pixels)
4. Corrected feature coordinates for rolling shutter: (adjusted reprojected observations - error)

5. Re-run bundle adjustment, triangulation or even entire reconstruction pipeline from scratch (not sure what method is best here, we'll experiment).

**Comment by pierotofy** [comment]
Corrected features (blue) vs. original features (red):

image

In this example, it compensates for the aircraft moving forward with a rolling shutter sensor, causing image dilation (so to counter dilation, we apply contraction).

**Comment by pierotofy** [comment]
Camera residuals without rolling shutter correction:

image

Vs. rolling shutter corrected:

image

**Comment by pierotofy** [comment]
3D RMS:

Without correction:

image

With correction:

image

**Comment by pierotofy** [comment]
This should be good to go :clinking_glasses:

File: opendm/rollingshutter.py (hunk 1 above)
Comment (pierotofy) [reply opendm/rollingshutter.py]: There we go: https://github.com/OpenDroneMap/RSCalibration

**Comment by Saijin-Naib** [comment]
Wow... Took out the final small remnants of bowling on my first dataset (3DR Solo + GoPro Hero4 Black), which was incredibly poorly collected:  
image

Reconstruction fidelity in general seems improved as well on smaller features.

**Pull request merged**